File: src/main/java/com/github/anviks/vixplugin/random_commands/protect_area/SerializableLocation.java
package com.github.anviks.vixplugin.random_commands.protect_area;

import kotlinx.serialization.Serializable;
import org.bukkit.Bukkit;
import org.bukkit.Location;

@Serializable
public class SerializableLocation {
    private final String world;
    private final int x;
    private final int y;
    private final int z;

    public SerializableLocation(Location location) {
        this.world = location.getWorld().getName();
        this.x = location.getBlockX();
        this.y = location.getBlockY();
        this.z = location.getBlockZ();
    }

    public SerializableLocation(String world, int x, int y, int z) {
        this.world = world;
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public String getWorld() {
        return world;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public int getZ() {
        return z;
    }

    public Location toLocation() {
        return new Location(
                Bukkit.getServer().getWorld(world),
                x,
                y,
                z
        );
    }
}

File: src/main/java/com/github/anviks/vixplugin/random_commands/protect_area/SerializableLocation.kt
package com.github.anviks.vixplugin.random_commands.protect_area

import kotlinx.serialization.Serializable
import org.bukkit.Bukkit
import org.bukkit.Location

@Serializable
class SerializableLocation {
    val world: String
    val x: Int
    val y: Int
    val z: Int

    constructor(location: Location) {
        this.world = location.world.name
        this.x = location.blockX
        this.y = location.blockY
        this.z = location.blockZ
    }

    constructor(world: String, x: Int, y: Int, z: Int) {
        this.world = world
        this.x = x
        this.y = y
        this.z = z
    }

    fun toLocation(): Location {
        return Location(
            Bukkit.getServer().getWorld(world),
            x.toDouble(),
            y.toDouble(),
            z.toDouble()
        )
    }
}
